File: conformance/reference-js/request-lifecycle.js
'use strict';

function applyRequestLifecycle(input) {
  let generation = input.initialGeneration ?? 0;
  let active = input.initialActive !== false;
  let acceptingResponses = generation > 0 && active;
  let state = input.initialState === undefined ? null : structuredClone(input.initialState);
  const committed = [];

  for (const event of input.events || []) {
    if (event.type === 'start') {
      generation += 1;
      active = true;
      acceptingResponses = true;
      continue;
    }
    if (event.type === 'hide' || event.type === 'unmount') {
      active = false;
      acceptingResponses = false;
      continue;
    }
    if (event.type === 'show') {
      active = true;
      continue;
    }
    if (event.type === 'response') {
      if (event.generation !== generation || !active || !acceptingResponses) continue;
      state = structuredClone(event.state);
      committed.push({ generation: event.generation, state: structuredClone(event.state) });
    }
  }

  return { generation, active, state, committed };
}

module.exports = { applyRequestLifecycle };

File: conformance/reference-python/request_lifecycle.py
def apply_request_lifecycle(input_value):
    generation = input_value.get("initialGeneration", 0)
    active = input_value.get("initialActive", True)
    accepting_responses = generation > 0 and active
    state = input_value.get("initialState")
    committed = []

    for event in input_value.get("events", []):
        event_type = event["type"]
        if event_type == "start":
            generation += 1
            active = True
            accepting_responses = True
        elif event_type in ("hide", "unmount"):
            active = False
            accepting_responses = False
        elif event_type == "show":
            active = True
        elif event_type == "response":
            if event["generation"] != generation or not active or not accepting_responses:
                continue
            state = event["state"]
            committed.append({"generation": event["generation"], "state": event["state"]})

    return {"generation": generation, "active": active, "state": state, "committed": committed}
